standard/expressions.md

```diff
@@ -6671,7 +6675,22 @@ The ref assignment operator shall not read the storage location referenced by th
 
 If the left operand of a compound assignment is of the form `E.P` or `E[Ei]` where `E` has the compile-time type `dynamic`, then the assignment is dynamically bound ([§12.3.3](expressions.md#1233-dynamic-binding)). In this case, the compile-time type of the assignment expression is `dynamic`, and the resolution described below will take place at run-time based on the run-time type of `E`. If the left operand is of the form `E[Ei]` where at least one element of `Ei` has the compile-time type `dynamic`, and the compile-time type of `E` is not an array, the resulting indexer access is dynamically bound, but with limited compile-time checking ([§12.6.5](expressions.md#1265-compile-time-checking-of-dynamic-member-invocation)).
 
-An operation of the form `x «op»= y` is processed by applying binary operator overload resolution ([§12.4.5](expressions.md#1245-binary-operator-overload-resolution)) as if the operation was written `x «op» y`. Then,
+For operator `??=`, given `a ??= b`, where `A` is the type of `a`, `B` is the type of `b`, and `A0` is the underlying type of `A`, if `A` is a nullable value type:
+
+1. If `A` does not exist or is a non-nullable value type, a compile-time error occurs.
+2. If `b` is not implicitly convertible to `A` or `A0` (if `A0` exists), a compile-time error occurs.
+3. If `A0` exists and `B` is implicitly convertible to `A0`, and `B` is not dynamic, then the type of `a ??= b` is `A0`. `a ??= b` is evaluated at runtime as:
+
+   ```csharp
+   var tmp = a.GetValueOrDefault();
+   if (!a.HasValue) { tmp = b; a = tmp; }
+   tmp
+   ```
+
+   Except that `a` is only evaluated once.
+4. Otherwise, the type of `a ??= b` is `A`. `a ??= b` is evaluated at runtime as `a ?? (a = b)`, except that `a` is only evaluated once.
```

Comment:
```suggestion
3. If `A0` exists and `b` is implicitly convertible to `A0`, and `B` is not dynamic, then the type of `a ??= b` is `A0`.
4. Otherwise, the type of `a ??= b` is `A`.

`a ??= b` is evaluated at runtime as `(T) (a ?? (a = b))`, where `T` is the type of the expression as determined above, except that `a` is only evaluated once.
```

Reply:
Stating “is evaluated at runtime as” makes it a requirement, elsewhere the phrases used are “is (semantically) equivalent to” and “has the same meaning as” and something similar should be used here. (But see seperate question following first.)

Reply:
Question: After the changes proposed to points 1 thru 4 do they all combine to:

> `a ??= b` is equivalent to `(T) (a ?? (a = b))` where `T` is the type of `a ?? b`.

Is this simply wrong, or is there some case this does not cover?

Reply:
The intention is that 1-4 now cover validity and expression type, so the evaluation can be covered separately. The complexity of 3 being evaluated in one way and 4 being evaluated in a different way felt unnecessary

Reply:
@jskeet I’m asking whether *all* of 1-4 can be removed and replaced by the suggestion, which I now see I missed “only once” from so it should be:

> `a ??= b` is equivalent to `(T) (a ?? (a = b))` where `T` is the type of `a ?? b`, except that `a` is evaluated only once.

Of something like that. This follows the other assignment operators in pushing the details to the existing non-assignment operator and equals, avoid repetition and having to keep two places in sync.

Except of course if the equivalence does not hold…

Reply:
Ah - the problem is that we need to establish `T` beforehand, and that's what 3 and 4 do... with the "well if we can be sure it's not null, the result type is not null" part being the bit that's slightly unusual.

Reply:
> Question: After the changes proposed to points 1 thru 4 do they all combine to:
> 
> > `a ??= b` is equivalent to `(T) (a ?? (a = b))` where `T` is the type of `a ?? b`.
> 
> Is this simply wrong, or is there some case this does not cover?

I think this is a great simplification that covers all cases except when `b` is dynamic.

```cs
void M(int? a, dynamic b)
{
    var x = a ?? b; // type: dynamic
    var y = a ??= b; // type: 'int?'
} 
```

So how about:
> `a ??= b` is equivalent to `(T) (a ?? (a = b))`, where `T` is the type of `a` when the type of `b` is dynamic and otherwise `T` is the type of `a ?? b`, except that `a` is evaluated only once.

Reply:
Again, resolving for clarity - but basically the plan is to use @jnm2's last suggestion.